### PR TITLE
add ssh command

### PR DESCRIPTION
**Pull request by empwilli**
This adds a new "ssh" command that makes use of the new ssh tunnel feature of the omnect-device-service. With this, one can use the CLI to create ssh tunnels to devices in the field. This is necessary, as sometimes devices may be in NATed networks and not directly reachable.

### CODE REVIEW

File: src/auth.rs [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+    let _ = open::that(auth_url.to_string());
+
+    let auth_code = server_task.await??;
+
+    let token_result = client
+        .exchange_code(AuthorizationCode::new(auth_code))
+        .set_pkce_verifier(pkce_verifier)
+        .request_async(async_http_client)
+        .await
+        .unwrap();
```
Comment (JanZachmann) [inline src/auth.rs:194]: why not `?` ?

File: src/lib.rs [hunk 2]
```diff
@@ -94,6 +98,34 @@ pub fn run() -> Result<()> {
...
+            ) -> Result<()> {
+                let access_token = crate::auth::authorize(&*crate::auth::AUTH_INFO_DEV)
+                    .await
+                    .context("create ssh tunnel")?;
+
+                let config = ssh::Config::new(dir, priv_key_path, config_path)?;
+
+                ssh::ssh_create_tunnel(device, username, config, access_token).await?;
+
+                Ok(())
```
Comment (JanZachmann) [inline src/lib.rs:124]: can be shortened by:?
`ssh::ssh_create_tunnel(device, username, config, access_token).await`

File: src/ssh.rs [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+use std::process::{Command, Stdio};
+use std::str;
+
+use log::warn;
+
+use anyhow::Result;
+use oauth2::AccessToken;
+use serde::{Deserialize, Serialize};
+
+static BACKEND_API_URL: &str = "https://cp.dev.omnect.conplement.cloud";
```
Comment (JanZachmann) [inline src/ssh.rs:14]: this must be configurable in order to support other tenants.

File: src/ssh.rs [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+        &mut writer,
+        r#"Host bastion
+	User {}
+	Hostname {}
+	Port {}
+	IdentityFile {}
+	CertificateFile {}
+	ProxyCommand none
+
+Host {}
```
Comment (JanZachmann) [inline src/ssh.rs:190]: indention

**Review by JanZachmann — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -163,6 +165,43 @@ Options:
   -b create bmap file
 ```
 
+# Creating an SSH Tunnel
+
+One can use `omnect-cli` to create a tunneled SSH connection to a device in the field. This is especially useful if the device is behind a NAT and can not directly be contacted. The device must have the `ssh` activated for this. Per default, this command will create a single use ssh key pair, certificate, and ssh configuration to establish a connection to the device.
+
+Note: if unused, the tunnel will close after 30 seconds.
```
[the remark this answers is not in the document]
Comment (empwilli) [reply README.md:172]: Probably. When testing, this time period was long enough for me to get a stable connection. Probably its best to discuss this when talking about the other security topics, as well.

**Review by JanZachmann — COMMENTED** [review]
still open questions

File: src/cli.rs [hunk 6]
```diff
@@ -149,8 +164,39 @@ pub enum IotHubDeviceUpdateConfig {
...
-/// This tools helps to manage your omnect devices. For more information visit:\nhttps://github.com/omnect/omnect-cli
+#[command(after_help = COPYRIGHT)]
+/// establish ssh connections to devices
+pub struct SshConfig {
+    /// username for the login on the device.
+    #[arg(short = 'u', long = "user", default_value = "omnect")]
+    pub username: String,
+    /// optional: path where the ssh key pair, the certificates, and the
+    /// temporary ssh configuration is stored. Defaults to
+    /// ${XDG_RUNTIME_DIR}/omnect-cli, or /tmp/omnect-cli, if the xdg variable
```
Comment (mlilien) [inline src/cli.rs:175]: this will only work for linux systems or wsl2 then, right? do we know how this behaves on macos?

File: src/util.rs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+use log::warn;
```
Comment (mlilien) [inline src/util.rs:1]: is this file referenced in more then one file?

File: Cargo.toml [hunk 8]
```diff
@@ -23,24 +23,36 @@ env_logger = "0.8"
 filemagic = "0.12"
 flate2 = "1.0.22"
 fs_extra = "1.2.0"
+futures = "0.3.28"
 futures-executor = "0.3.15"
 futures-util = "0.3.15"
 hex = "0.4.3"
+http = "0.2.9"
+http-body-util = "0.1.0-rc.2"
+hyper = { version = "1.0.0-rc.3", features = ["http1", "server"] }
```
[the remark this answers is not in the document]
Comment (empwilli) [reply Cargo.toml]: replaced w/ actix.

File: Cargo.toml [hunk 9]
```diff
@@ -23,24 +23,36 @@ env_logger = "0.8"
...
 lazy_static = "1.4.0"
 log = "0.4"
 num_cpus = "1.13"
+oauth2 = "4.4.1"
+open = "4.1.0"
 openssl = { version = "0.10" }
 openssl-sys = "0.9"
 path-absolutize = "3.0.10"
 regex = "1.5.5"
+reqwest = { version = "0.11.18", features = ["json"] }
```
[the remark this answers is not in the document]
Comment (empwilli) [reply Cargo.toml:43]: oauth2 already ships reqwest; so this is already a dependency; we could use awc alternatively, but the code does not change significantly at the one call side.

File: src/cli.rs (hunk 6 above)
Comment (empwilli) [reply src/cli.rs]: dropped

File: src/util.rs (hunk 7 above)
Comment (empwilli) [reply src/util.rs]: dropped

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (empwilli) [reply README.md]: Updated to 5 minutes.

File: README.md [hunk 10]
```diff
@@ -10,12 +10,15 @@ omnect-cli is a cli tool to manage your omnect-devices. It provides commands to
   - Inject an iotedge gateway identity configuration for AIS
   - Inject an iot leaf identity configuration for AIS
   - Inject a device certificate with corresponding key from a given intermediate full-chain-certificate and corresponding key
+  - Inject an ssh root ca and device principal for ssh tunnel creation.
```
[the remark this answers is not in the document]
Comment (empwilli) [reply README.md:13]: SSH is pronounced with an open vowel, as such "an" is more appropriate here (https://ell.stackexchange.com/a/53247)

**Review by JanZachmann — APPROVED** [review]

**Pull request merged**